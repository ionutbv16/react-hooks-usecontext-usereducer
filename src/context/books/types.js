export const GET_BOOKS = "GET_BOOKS";
export const SET_LOADING = "SET_LOADING";
