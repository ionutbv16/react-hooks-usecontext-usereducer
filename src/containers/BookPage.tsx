import React, { useContext } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import styled from 'styled-components';
import { Button } from '@material-ui/core';
import BooksContext from '../context/books/booksContext'
import { Title, Wrapper } from '../commonstyled/styled';
import BookDetails from './components/BookDetails';
import BookItem from './components/BookItem';
import { ParamTypes } from '../types';

const ButtonDivider = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding-top: 20px;
  padding-bottom: 30px;
`;

const BooksWrapper = styled.div`
  border: 1px solid rgba(0, 0, 0, 0.12);
  border-radius: 8px;
  color: rgba(0, 0, 0, 0.87); 
  background-color: #fff;
  padding:20px;
`;

const BooksContainer = styled.div`
`;

const BookPage: React.FC = () => {

  const history = useHistory();
  let params = useParams<ParamTypes>();
  const booksContext = useContext(BooksContext);
  const { loading, books, getBooks } : any = booksContext

  const  booksResults :any = books.books

  if (!booksResults) {
    history.push("/")
    return null
  }
  else {
      const [book] =  (booksResults && booksResults.length > 0) && booksResults.filter( (element: any) => element.id === params.id )
      const { title, author, cover, isbn } = book
      const sameAuthor = (booksResults && booksResults.length > 0) && booksResults.filter( (element: any) => (element.author == author && element.id !== params.id ) )
    
      const renderCards = () => {
        if (sameAuthor && sameAuthor.length > 0) {
           return sameAuthor.map((item:any) => <BookItem history={history}  id={item.id} title={item.title} author={item.author} cover={item.cover} />);
        }
      };

      return (
        <BooksContainer>
          <Wrapper>
            <BooksWrapper>
              <Title>Book Details</Title>
                <BookDetails  title={title} author={author} cover={cover} isbn={isbn} />
              </BooksWrapper>

              <ButtonDivider>
                <Button variant="contained" onClick={() => history.push("/")}>Back  Home</Button>
              </ButtonDivider>
          </Wrapper>

          { sameAuthor && sameAuthor.length > 0  && (
            <Wrapper>
              <BooksWrapper>
              { renderCards()}
              </BooksWrapper>
            </Wrapper>
          )}
        </BooksContainer>
      );
    }
};

export default BookPage
