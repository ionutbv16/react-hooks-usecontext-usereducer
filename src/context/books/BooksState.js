import React, { useReducer } from "react";
import axios from "axios";
import BooksContext from "./booksContext";
import BooksReducer from "./booksReducer";
import { GET_BOOKS, SET_LOADING } from "./types";
import { URL_FETCH } from "../../config";

const BooksState = (props) => {
  const initialState = {
    books: [],
    loading: false,
  };

  const [state, dispatch] = useReducer(BooksReducer, initialState);

  const getBooks = async () => {
    setLoading();
    const res = await axios.get(URL_FETCH);
    dispatch({
      type: GET_BOOKS,
      payload: res.data,
    });
  };

  const setLoading = () => dispatch({ type: SET_LOADING });

  return (
    <BooksContext.Provider
      value={{
        books: state.books,
        loading: state.loading,
        getBooks,
      }}
    >
      {props.children}
    </BooksContext.Provider>
  );
};

export default BooksState;
