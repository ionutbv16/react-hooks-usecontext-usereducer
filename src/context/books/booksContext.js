import { createContext, useContext } from "react";

const booksContext = createContext();
export const useBooksContext = () => useContext(booksContext);

export default booksContext;
