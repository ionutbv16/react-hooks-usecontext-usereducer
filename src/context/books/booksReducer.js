import { GET_BOOKS, SET_LOADING } from "./types";

export default (state, action) => {
  switch (action.type) {
    case GET_BOOKS: {
      return {
        ...state,
        books: action.payload,
        loading: false,
      };
    }
    case SET_LOADING:
      return {
        ...state,
        loading: true,
      };
    default:
      return state;
  }
};
