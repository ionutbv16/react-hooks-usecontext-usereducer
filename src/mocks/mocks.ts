 
export const booksMock = {"books":[{"id":"872179f2-4de2-4cde-a259-ee470d83d515","cover":"https://lorempixel.com/640/480/?ee470d83d515","isbn":"9781593275846","title":"Eloquent JavaScript, Second Edition","subtitle":"A Modern Introduction to Programming","author":"Mrs. John Doe","published":"2014-12-14T00:00:00.000Z","publisher":"No Starch Press","pages":472,"description":"JavaScript lies at the heart of almost every modern web application, from social apps to the newest browser-based games. Though simple for beginners to pick up and play with, JavaScript is a flexible, complex language that you can use to build full-scale applications.","website":"http://eloquentjavascript.net/"}]};
 
