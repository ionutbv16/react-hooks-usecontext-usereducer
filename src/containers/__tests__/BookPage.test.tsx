import React from 'react';
import BookPage from '../BookPage'; 
import { shallow } from 'enzyme'
import * as BooksContextModule from '../../context/books/booksContext'
import { booksMock } from '../../mocks/mocks'
import { MemoryRouter } from 'react-router';

describe('Test BookPage Component', () => {
  test('should render without error while mocking useContext', () => {
    const contextValues = booksMock
    jest
      .spyOn(BooksContextModule, 'useBooksContext')
      .mockImplementation(() => contextValues);
    const wrapper = shallow(<MemoryRouter><BookPage /></MemoryRouter>);
    expect(wrapper).toMatchSnapshot();
  });
});