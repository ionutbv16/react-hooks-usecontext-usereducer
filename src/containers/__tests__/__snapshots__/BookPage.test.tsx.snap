// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test BookPage Component should render without error while mocking useContext 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "isb450",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "isb450",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <BookPage />
</Router>
`;
